### PR TITLE
Improve the configuration of the running gem

### PR DESCRIPTION
**Pull request by malakai97**
Relatively minor change, as the values are still available off of `Fauxpaas.whatever`, as before.  Better description is in the [gist](https://gist.github.com/malakai97/affbdb0902c244c9d8ca3fe72b08d6cc).

### CODE REVIEW

File: lib/fauxpaas/cli/main.rb [hunk 1]
```diff
@@ -15,13 +17,15 @@ class Main < Thor
       desc "deploy <instance>",
         "Deploys the instance's source; by default deploys master. Use --reference to deploy a specific revision"
       def deploy(instance_name)
+        Fauxpaas.config = configuration(options)
```
[the remark this answers is not in the document]
Comment (malakai97) [reply lib/fauxpaas/cli/main.rb]: This is here because the CLI modifies the configuration with options it gets from the command line. Ideally there would be some sort of hook after option parsing that we could attach this to in Thor, but I could not find one.

File: lib/fauxpaas/cli/main.rb [hunk 2]
```diff
@@ -41,16 +45,35 @@ def default_branch(instance_name, new_branch = nil)
       desc "rollback <instance> [<cache>]",
         "Rollsback to the specified cache, or the most recent one."
       def rollback(instance_name)
-        instance = Fauxpaas.instance_repo.find(instance_name)
+        Fauxpaas.config = configuration(options)
+        instance = Fauxpaas.config.instance_repo.find(instance_name)
```
[the remark this answers is not in the document]
Comment (malakai97) [reply lib/fauxpaas/cli/main.rb]: > Fauxpaas, by using method_missing is sort of having an identity crisis... Is it a singleton component or is it just a module for some settings and namespacing?

fauxpaas/configuration.rb extends the module by adding configuration to it, including the delegation pattern. One of the more brittle patterns can be used when the need arises.

> I'm okay with Fauxpaas serving as a singleton. In that case, I would probably put a configure method on it that takes options (to bootstrap itself) and expose its collaborators (instance_repo and deployer) explicitly.

I don't see what we gain by doing that. It does not seem functionally different from inlining the Configuration class into the Fauxpaas module and changing the `def =` to `def configure`.

File: lib/fauxpaas/configuration.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    @config = value
+  end
+
+  def self.config
+    @config ||= Configuration.new
+  end
+
+  # Avoid having to type Fauxpaas.config.thing_i_want by
+  # delegating here.
+  def self.method_missing(method, *args)
```
[the remark this answers is not in the document]
Comment (malakai97) [reply lib/fauxpaas/configuration.rb]: The module has no functionality other than to make the config accessible globally. Every method delegates here.

File: lib/fauxpaas/configuration.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    end
+    def gem_root
+      @lib_root ||= Pathname.new(__FILE__).dirname.parent.parent
+    end
+
+    def instance_root
+      @instance_root ||= gem_root + "deploy" + "instances"
+    end
+
+    def instance_repo
```
[the remark this answers is not in the document]
Comment (malakai97) [reply lib/fauxpaas/configuration.rb]: I get the impression that you want "configuration" to mean a configuration file.  A DI container _is_ a configuration.  The service that gets returned by a given query _is_ determined by the configuration.

If we were to use something like the railsconfig gem, wouldn't we be doing the exact same thing? Read some stuff from files, apply some defaults, and make the resulting objects available globally.

**Comment by malakai97** [comment]
tl;dr: I don't like your suggestions :face_with_head_bandage: 

I could maybe be convinced if you showed me what you wanted this to look like ideally. It seemed like you were trying to change my solution to make you happy, but ended up with one neither of us liked.
